superset-frontend/src/explore/components/DatasourcePanel/index.tsx

```diff
@@ -203,45 +207,147 @@ export default function DataSourcePanel({
             header={<span className="header">{t('Metrics')}</span>}
             key="metrics"
           >
-            <div className="field-length">
-              {t(`Showing %s`, lists.metrics.length)}
-            </div>
-            {lists.metrics.map(m => (
-              <LabelContainer key={m.metric_name} className="column">
-                {enableExploreDnd ? (
-                  <DatasourcePanelDragWrapper
-                    value={m}
-                    type={DndItemType.Metric}
-                  >
-                    <MetricOption metric={m} showType />
-                  </DatasourcePanelDragWrapper>
-                ) : (
-                  <MetricOption metric={m} showType />
-                )}
-              </LabelContainer>
-            ))}
+            { lists.metrics.length > 50 ?
+              <>
+                { showAll ?
+                  <>
+                    <div className="field-length">
+                      {t(`Showing %s`, lists.metrics.length)}
+                    </div>
+                    {lists.metrics.map(m => (
+                      <LabelContainer key={m.metric_name} className="column">
+                        {enableExploreDnd ? (
+                          <DatasourcePanelDragWrapper
+                            value={m}
+                            type={DndItemType.Metric}
+                          >
+                            <MetricOption metric={m} showType />
+                          </DatasourcePanelDragWrapper>
+                        ) : (
+                          <MetricOption metric={m} showType />
+                        )}
+                      </LabelContainer>
+                    ))}
+                  </>
+                  :
+                  <>
+                    <div className="field-length">
+                      {t(`Showing %s of %s`, metricSlice.length, lists.metrics.length)}
+                    </div>
+                    {metricSlice.map(m => (
+                      <LabelContainer key={m.metric_name} className="column">
```

Comment:
Open question for folks who know React better than I do: is it possible to reduce the code duplication here in a clean way?

Reply:
Now that you bring it up, I could probably just create a separate function that contains all that code and call it when needed. Not sure why I didn't think of that on Friday.